clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java

```diff
@@ -354,14 +373,18 @@ public Cluster cluster() {
             if (metadata.error == Errors.NONE) {
                 if (metadata.isInternal)
                     internalTopics.add(metadata.topic);
-                for (PartitionMetadata partitionMetadata : metadata.partitionMetadata)
+                for (PartitionMetadata partitionMetadata : metadata.partitionMetadata) {
+                    Integer leaderEpoch = partitionMetadata.leaderEpoch == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
```

Comment:
Yeah, sorry about that. I was struggling to find a consistent pattern given the complication of `OffsetAndMetadata` and the apparently frowned-upon use of `Optional` in fields. It is helpful generally when argument types match return types since it makes composition easier, but I'm not sure how you can achieve that with the standard usage advice for `Optional`. I think our options are the following:

1. Use `Optional<Integer>` everywhere (including in fields) except within `OffsetAndMetadata`.
2. Use nullable integer internally, but expose `Optional<Integer>` from public APIs.
3. Use primitive int with a sentinel internally, but expose `Optional<Integer>` from public APIs.

I was trying hard to avoid the third option because unsafe use of sentinel values can cause significantly more damage than unsafe use of an optional or nullable value. At least in the latter case, you get a fairly clear exception which is raised from the point of usage. Improper use of a sentinel value can have a massive impact at a distance from the usage (as mentioned above).

My inclination at the moment is probably option 1, but I'd like to hear other opinions.

Reply:
I prefer `1` which is the approach we use in Scala and I believe has worked out pretty well. It's annoying that Java came up with its own weird guidelines about `Optional` usage. One concrete outcome is that it doesn't implement `Serializable`. Since we only implement `Serializable` in a couple of classes, this doesn't seem like a blocker.

Reply:
Sounds good. Let's use option 1.